### PR TITLE
fix(language-service): use single entry point for Ivy and View Engine

### PR DESCRIPTION
**Pull request by kyliau**
Currently there are two entry points for the `@angular/language-service`
package:

- `@angular/language-service`
  This default entry point is for View Engine LS. Through the redirection
  of `main` field in `package.json`, it resolves to
  `./bundles/language-service.js`.
- `@angular/language-service/bundles/ivy.js`
  This secondary entry point is for Ivy LS.

TypeScript recently changed the behavior of tsserver to allow only package
names as plugin names [1] for security reasons. This means the secondary
entry point for Ivy LS can no longer be used.
We implemented a quick [hack](https://github.com/angular/vscode-ng-language-service/pull/1111) in the module resolver (in the extension repo)
to fix this, but the long term fix should be in `@angular/language-service`.

Here, the `main` field in `package.json` is changed to `index.js`, and in the
index file we conditionally load View Engine or Ivy based on the input config.
This eliminates the need for multiple entry points.

As part of this PR, I also removed all source code for View Engine and Ivy
included in the NPM package. Consumers of this package should run the bundled
output and nothing else. This would help us prevent an accidental import that
results in execution of unbundled code.

The new layout of the NPM package is as follows:
```
@angular/language-service
├── api.d.ts
├── api.js
├── bundles
│   ├── ivy.js
│   └── language-service.js
├── index.d.ts
├── index.js
└── package.json
```

[1]: https://github.com/microsoft/TypeScript/pull/42713

cc: @ivanwonder 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: packages/language-service/index.ts [hunk 1]
```diff
@@ -6,5 +6,21 @@
...
+  plugin = require(`./bundles/${bundleName}`);
+  return plugin.create(info);
+}
+
+export function getExternalFiles(proj: ts.server.Project): string[] {
+  return plugin?.getExternalFiles?.(proj) || [];
+}
+
+export function onConfigurationChanged(config: any): void {
+  plugin?.onConfigurationChanged?.(config);
```
Comment (atscott) [inline packages/language-service/index.ts:25]: Hmm, this is a _little_ bit awkward given that you have to call `create` first but I can't think of a better way...

File: packages/language-service/index.ts [hunk 2]
```diff
@@ -6,5 +6,21 @@
...
+let plugin: ts.server.PluginModule;
+
+export function create(info: ts.server.PluginCreateInfo): ts.LanguageService {
+  const bundleName = info.config.ivy ? 'ivy.js' : 'language-service.js';
+  plugin = require(`./bundles/${bundleName}`);
+  return plugin.create(info);
+}
+
+export function getExternalFiles(proj: ts.server.Project): string[] {
+  return plugin?.getExternalFiles?.(proj) || [];
```
Comment (atscott) [inline packages/language-service/index.ts:21]:
```suggestion
  return plugin?.getExternalFiles?.(proj) ?? [];
```
nit: nullish coalesce

File: packages/language-service/index.ts [hunk 3]
```diff
@@ -6,5 +6,21 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import * as ts from 'typescript/lib/tsserverlibrary';
 export * from './api';
-export {create, getExternalFiles} from './src/ts_plugin';
+
+let plugin: ts.server.PluginModule;
+
+export function create(info: ts.server.PluginCreateInfo): ts.LanguageService {
```
Comment (atscott) [inline packages/language-service/index.ts:14]: Should we return `NgLanguageService` here? It likely doesn't matter, but I suppose it would be technically more accurate, right?

**Review by atscott — APPROVED** [review]

**Comment by atscott** [comment]
Quick question @kyliau  - would this be considered a breaking change since it changes the "main" property in `package.json`? Or is there another reason this targets `major` only?

**Comment by ivanwonder** [comment]
https://github.com/angular/angular/blob/a6971ba89adc253bfa4260036ee4a1e0bd76159f/packages/language-service/bundles/BUILD.bazel#L6
@kyliau the `entry_point` is wrong for ve bundles.

File: packages/language-service/index.ts (hunk 1 above)
Comment (ivanwonder) [reply packages/language-service/index.ts]: Whether we can assert the `plugin` has been created by tsserver. If not, the tsserver cannot invoke the function.

https://github.com/microsoft/TypeScript/blob/5280ba400c071c87e6ba25c7d8d1ac9c304fe2d4/src/server/project.ts#L1645-L1651

File: packages/language-service/index.ts [hunk 4]
```diff
@@ -6,5 +6,39 @@
...
+
 export * from './api';
-export {create, getExternalFiles} from './src/ts_plugin';
+
+interface PluginModule extends ts.server.PluginModule {
+  create(createInfo: ts.server.PluginCreateInfo): NgLanguageService;
+  onConfigurationChanged?(config: NgLanguageServiceConfig): void;
+}
+
+const factory: ts.server.PluginModuleFactory = function(mod): PluginModule {
```
Comment (gkalpak) [inline packages/language-service/index.ts:19]: OOC, why use a regular function instead of an arrow function?

File: packages/language-service/BUILD.bazel [hunk 5]
```diff
@@ -6,15 +6,17 @@ ts_library(
     name = "api",
     srcs = [
         "api.ts",
+        "index.ts",
     ],
     deps = [
+        "@npm//@types/node",
```
Comment (gkalpak) [inline packages/language-service/BUILD.bazel:12]: Why is this needed now (and it wasn't before)?

**Review by gkalpak — APPROVED** [review]
LGTM :+1:
(Just a couple of question - out of curiosity mostly 🧐)

Reviewed-for: dev-infra

File: packages/language-service/index.ts (hunk 4 above)
Comment (kyliau) [reply packages/language-service/index.ts:19]: Google3 style guide usually recommends using function foo() for function declaration, but in this case you are right:
> Top level arrow functions may be used to explicitly declare that a function implements an interface.

I've changed this to arrow function.

(internal [link](https://g3doc.corp.google.com/javascript/typescript/g3doc/styleguide.md?cl=head#function-declarations) for reference)

File: packages/language-service/BUILD.bazel (hunk 5 above)
Comment (kyliau) [reply packages/language-service/BUILD.bazel:12]: because the implementation uses `require`

File: packages/language-service/index.ts (hunk 3 above)
Comment (kyliau) [reply packages/language-service/index.ts]: done

**Comment by kyliau** [comment]
> Quick question @kyliau - would this be considered a breaking change since it changes the "main" property in `package.json`? Or is there another reason this targets `major` only?


@atscott It's technically not a breaking change since `@angular/language-service` is considered a "private" package, much like `@angular/compiler-cli`. But since I'm making drastic changes to the package contents (removed all source files), I don't feel comfortable doing it in a patch release.

**Comment by kyliau** [comment]
> https://github.com/angular/angular/blob/a6971ba89adc253bfa4260036ee4a1e0bd76159f/packages/language-service/bundles/BUILD.bazel#L6
> 
> 
> @kyliau the `entry_point` is wrong for ve bundles.

@ivanwonder Thanks for pointing that out, I've fixed the entry point :)
